src/cdi_CPEloss/Tabular_CP_Eloss.cc

```diff
@@ -0,0 +1,270 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!                                                                              
+ * \file   cdi_CPEloss/Tabular_CP_Eloss.cc                                     
+ * \author Ben R. Ryan                                                        
+ * \date   2019 Nov 4                                             
+ * \brief  Tabular_CP_Eloss member definitions.                                 
+ * \note   Copyright (C) 2016-2019 Triad National Security, LLC.                 
+ *         All rights reserved. */
+//---------------------------------------------------------------------------//
+
+#include "Tabular_CP_Eloss.hh"
+
+namespace rtt_cdi_cpeloss {
+
+//---------------------------------------------------------------------------//
+// CONSTRUCTORS
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Constructor for an analytic tabular eloss model. 
+ *
+ * This constructor builds an eloss model defined by the
+ * rtt_cdi_cpeloss::Tabular_Eloss_Model derived class argument.
+ *
+ * The path to an eloss datafile is passed to the constructor,
+ * which opens and parses the file. The file format is the usual
+ * LANL format for stopping powers. 
+ *
+ * \param[in] filename_in path to eloss file
+ * \param[in] target_in target particle zaid
+ * \param[in] projectile_in transporting particle zaid
+ */
+Tabular_CP_Eloss::Tabular_CP_Eloss(std::string filename_in,
+                                   rtt_cdi::CParticle target_in,
+                                   rtt_cdi::CParticle projectile_in)
+    : rtt_cdi::CPEloss(target_in, projectile_in), filename(filename_in) {
+  using std::stod;
+  using std::stoi;
+
+  model_type = rtt_cdi::CPModelType::TABULAR_ETYPE;
+
+  file.open(filename);
+  std::ostringstream file_error;
+  file_error << "Error opening DEDX file \"" << filename << "\"";
+  Insist(file.is_open(), file_error.str());
+
+  std::vector<std::string> line_entries;
+  uint32_t nlines;
+  constexpr uint32_t max_entries =
+      6; // This is a statement about the file format, maximum of six entries per row.
+
+  line_entries = read_line(); // ZAID
+  int32_t projectile_zaid_file = stoi(line_entries[0]);
+  Require(projectile.get_zaid() == projectile_zaid_file);
+
+  read_line(); // Z, A, mass
+
+  line_entries = read_line(); // Number of bins for energy, density, temperature
+  n_energy = stoi(line_entries[0]);
```

Comment:
Consider adding `Check(line_entries.size() >= 3)` before accessing these elements.

Reply:
Added a similar check for the bin spacing line.